### PR TITLE
Deserialize for std::collections

### PR DESCRIPTION
**Pull request by naomijub**
adds `#[deprecated]` to `ser_struct!`

### CODE REVIEW

File: src/deserialize/mod.rs [hunk 1]
```diff
@@ -177,6 +186,128 @@ where
...
+impl<T: std::cmp::Eq + std::hash::Hash> Deserialize for HashSet<T>
+where
+    T: Deserialize,
+{
+    fn deserialize(edn: &Edn) -> Result<Self, Error> {
+        match edn {
+            Edn::Set(_) => Ok(edn
+                .set_iter()
+                .ok_or_else(|| Error::Iter(format!("Could not create iter from {:?}", edn)))?
+                .map(|e| Deserialize::deserialize(e).unwrap())
```
Comment (evaporei) [inline src/deserialize/mod.rs:276]: I think we should favor `from_edn` instead of `Deserialize::deserialize`

Comment (evaporei) [reply src/deserialize/mod.rs:276]: Also, shouldn't we return the possible errors too? This seems unsafe

File: src/deserialize/mod.rs [hunk 2]
```diff
@@ -177,6 +186,128 @@ where
...
+impl<T> Deserialize for BTreeMap<String, T>
+where
+    T: Deserialize,
+{
+    fn deserialize(edn: &Edn) -> Result<Self, Error> {
+        match edn {
+            Edn::Map(_) => Ok(edn
+                .map_iter()
+                .ok_or_else(|| Error::Iter(format!("Could not create iter from {:?}", edn)))?
+                .map(|(key, e)| (key, Deserialize::deserialize(e).unwrap()))
```
Comment (evaporei) [inline src/deserialize/mod.rs:237]: Again, `from_edn` > `Deserialize::deserialize` and also unsafety of using `unwrap`.

File: src/deserialize/mod.rs [hunk 3]
```diff
@@ -177,6 +186,128 @@ where
...
+                .map_iter()
+                .ok_or_else(|| Error::Iter(format!("Could not create iter from {:?}", edn)))?
+                .map(|(key, e)| {
+                    (
+                        if ns.starts_with(":") {
+                            ns.to_string() + "/" + key
+                        } else {
+                            String::from(":") + ns + "/" + key
+                        },
+                        Deserialize::deserialize(e).unwrap(),
```
Comment (evaporei) [inline src/deserialize/mod.rs:252]: `from_edn` and `unwrap`

File: src/deserialize/mod.rs [hunk 4]
```diff
@@ -177,6 +186,128 @@ where
...
+impl<T> Deserialize for HashMap<String, T>
+where
+    T: Deserialize,
+{
+    fn deserialize(edn: &Edn) -> Result<Self, Error> {
+        match edn {
+            Edn::Map(_) => Ok(edn
+                .map_iter()
+                .ok_or_else(|| Error::Iter(format!("Could not create iter from {:?}", edn)))?
+                .map(|(key, e)| (key, Deserialize::deserialize(e).unwrap()))
```
Comment (evaporei) [inline src/deserialize/mod.rs:198]: `from_edn` and `unwrap`

File: src/deserialize/mod.rs [hunk 5]
```diff
@@ -177,6 +186,128 @@ where
...
+                .map_iter()
+                .ok_or_else(|| Error::Iter(format!("Could not create iter from {:?}", edn)))?
+                .map(|(key, e)| {
+                    (
+                        if ns.starts_with(":") {
+                            ns.to_string() + "/" + key
+                        } else {
+                            String::from(":") + ns + "/" + key
+                        },
+                        Deserialize::deserialize(e).unwrap(),
```
Comment (evaporei) [inline src/deserialize/mod.rs:213]: `from_edn` and `unwrap`

File: src/deserialize/mod.rs (hunk 1 above)
Comment (naomijub) [reply src/deserialize/mod.rs:276]: every other deserialize uses `Deserialize::deserialize` and I think `from_edn`is something more usable from the client side, as well as `Deserialize::deserialize` is more explicit to have in an internal library context

Comment (naomijub) [reply src/deserialize/mod.rs:276]: couldn't resolve `unwrap`

**Comment by naomijub** [comment]
https://github.com/naomijub/edn-rs/issues/68

**Pull request merged**